Realm/RLMQueryUtil.mm

```diff
@@ -113,6 +141,68 @@ void set_base_table(const Table*) override {}
     }
 };
 
+
+// Equal and ContainsSubstring are used by QueryBuilder::add_string_constraint as the comparator
+// for performing diacritic-insensitive comparisons.
+
+template <NSUInteger options>
+struct Equal {
+    using CaseSensitive = Equal<options & ~NSCaseInsensitiveSearch>;
+    using CaseInsensitive = Equal<options | NSCaseInsensitiveSearch>;
+
+    bool operator()(StringData v1, StringData v2, bool v1_null, bool v2_null) const
+    {
+        REALM_ASSERT_DEBUG(v1_null == v1.is_null());
+        REALM_ASSERT_DEBUG(v2_null == v2.is_null());
+
+        if (v1_null || v2_null) {
+            return v1_null == v2_null;
+        }
+
+        @autoreleasepool {
+            NSString *s1 = [[NSString alloc] initWithBytesNoCopy:(void*)v1.data() length:v1.size() encoding:NSUTF8StringEncoding freeWhenDone:NO];
```

Comment:
I had compared the two back when I first started looking at it, and the differences appeared to be within the noise (a rather large amount of noise at the time). I just compared again with a better test scenario (larger table, more iterations of the test) and there's a measurable ~15% win from using the CF APIs.